lib/backend.js

```diff
@@ -252,15 +253,59 @@ Backend.prototype._sanitizeSnapshots = function(agent, projection, collection, s
     }
   }
 
-  var request = {
-    collection: collection,
-    snapshots: snapshots,
-    snapshotType: snapshotType
-  };
+  var request = new ReadSnapshotsRequest(collection, snapshots, snapshotType);
 
-  this.trigger(this.MIDDLEWARE_ACTIONS.readSnapshots, agent, request, callback);
+  this.trigger(this.MIDDLEWARE_ACTIONS.readSnapshots, agent, request, function(err) {
+    if (err) return callback(err);
+    // Handle "partial rejection" - "readSnapshots" middleware functions can use
+    // `request.rejectSnapshotRead(snapshot, error)` to reject the read of a specific snapshot.
+    if (request.hasSnapshotRejection()) {
+      err = getReadSnapshotsError(snapshots, request._idToError);
+    }
+    if (err) {
+      callback(err);
+    } else {
+      callback();
+    }
+  });
 };
 
+/**
+ * Returns an overall error from "readSnapshots" based on the snapshot-specific errors.
+ *
+ * - If there's exactly one snapshot and it has an error, then that error is returned.
+ * - If there's more than one snapshot and at least one has an error, then an overall
+ *   "ERR_SNAPSHOT_READS_REJECTED" is returned, with an `idToError` property.
+ *
+ * @param {Snapshot[]} snapshots
+ * @param {{[docId: string]: string | Error}} idToError
+ */
+function getReadSnapshotsError(snapshots, idToError) {
```

Comment:
I think this function should be moved into ReadSnapshotsRequest. It needs to access the id map, which should really be private to the class.